File: server/middleware/rate-limit.ts
import { defineEventHandler, setResponseHeader, createError } from "h3";
import { getRedisClient } from "../utils/redis";

// Rate limit config (mirrors the old express-rate-limit setup)
const LIMITS = {
  global: { requests: 200, windowMs: 15 * 60 * 1000 }, // 200 req / 15 min
} as const;

// Lightweight sliding window rate limiter using Redis INCR + EXPIRE
async function checkRateLimit(
  key: string,
  limit: number,
  windowMs: number
): Promise<{ allowed: boolean; remaining: number }> {
  const redis = getRedisClient();
  const windowSec = Math.ceil(windowMs / 1000);

  const current = await redis.incr(key);
  if (current === 1) {
    await redis.expire(key, windowSec);
  }

  const allowed = current <= limit;
  const remaining = Math.max(0, limit - current);
  return { allowed, remaining };
}

export default defineEventHandler(async (event) => {
  const url = event.node.req.url ?? "";

  // Only rate-limit API routes
  if (!url.startsWith("/api/")) return;

  const ip =
    (event.node.req.headers["x-forwarded-for"] as string)?.split(",")[0]?.trim() ||
    event.node.req.socket?.remoteAddress ||
    "unknown";

  const key = `ratelimit:global:${ip}`;

  try {
    const { allowed, remaining } = await checkRateLimit(
      key,
      LIMITS.global.requests,
      LIMITS.global.windowMs
    );

    setResponseHeader(event, "X-RateLimit-Limit", String(LIMITS.global.requests));
    setResponseHeader(event, "X-RateLimit-Remaining", String(remaining));

    if (!allowed) {
      throw createError({
        statusCode: 429,
        message: "Too many requests, please slow down.",
      });
    }
  } catch (err: any) {
    // If Redis is unavailable, fail open (don't block traffic)
    if (err.statusCode === 429) throw err;
    console.error("[rate-limit] Redis error, failing open:", err.message);
  }
});

File: actions/rate_limit.go
package actions

import (
	"log"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

// RateLimiterMiddleware limits requests to API endpoints to 200 req / 15 min using Redis
func RateLimiterMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		req := c.Request
		url := req.URL.Path

		// Only rate-limit API routes
		if !strings.HasPrefix(url, "/api/") {
			c.Next()
			return
		}

		// Get IP address
		ip := c.GetHeader("X-Forwarded-For")
		if ip != "" {
			ip = strings.Split(ip, ",")[0]
			ip = strings.TrimSpace(ip)
		} else {
			var err error
			ip, _, err = net.SplitHostPort(req.RemoteAddr)
			if err != nil {
				ip = req.RemoteAddr
			}
		}

		key := "ratelimit:global:" + ip
		limit := int64(200)
		window := 15 * time.Minute

		allowed := true
		remaining := limit

		if RedisClient != nil {
			ctx := req.Context()
			current, err := RedisClient.Incr(ctx, key).Result()
			if err == nil {
				if current == 1 {
					RedisClient.Expire(ctx, key, window)
				}
				allowed = current <= limit
				if limit-current > 0 {
					remaining = limit - current
				} else {
					remaining = 0
				}
			} else {
				log.Printf("[rate-limit] Redis error, failing open: %s", err)
			}
		}

		// Set rate-limit headers
		c.Header("X-RateLimit-Limit", strconv.FormatInt(limit, 10))
		c.Header("X-RateLimit-Remaining", strconv.FormatInt(remaining, 10))

		if !allowed {
			c.JSON(http.StatusTooManyRequests, gin.H{
				"error": "Too many requests, please slow down.",
			})
			c.Abort()
			return
		}

		c.Next()
	}
}
